src/shared/dissect-image.c

```diff
@@ -2639,16 +2639,19 @@ static int do_crypt_activate_verity(
                 if (r >= 0)
                         return r;
 
-                log_debug("Validation of dm-verity signature failed via the kernel, trying userspace validation instead.");
+                log_debug_errno(r, "Validation of dm-verity signature failed via the kernel, trying userspace validation instead: %m");
 #else
                 log_debug("Activation of verity device with signature requested, but not supported via the kernel by %s due to missing crypt_activate_by_signed_key(), trying userspace validation instead.",
                           program_invocation_short_name);
+                r = 0; /* Set for RET_GATHER() below */
 #endif
 
                 /* So this didn't work via the kernel, then let's try userspace validation instead. If that
                  * works we'll try to activate without telling the kernel the signature. */
 
-                r = validate_signature_userspace(verity);
+                /* Propagate the original kernel error, so that the fallback logic can work, as the
+                 * device-mapper is finicky around concurrent activations of the same volume */
+                RET_GATHER(r, validate_signature_userspace(verity));
```

Comment:
hmm, so let's say kernel has no key for this, but userspace has. with your new change we'd fail in that case, but we should allow things.

i.e. logic should probably be:

```c
r = try_kernel;
if (r >= 0)
        return r;

k = validate_signature_userspace();
if (k < 0)
        return r < 0 ? r : k; /* preferably return kernel error */
if (k == 0)
        return log_debug_errno(r /* ditto */, "Activation …");
```

Reply:
ok, updated